web/pandas/pdeps/0005-no-default-index-mode.md

```diff
@@ -0,0 +1,189 @@
+# PDEP-5: No-default-index mode
+
+- Created: 14 November 2022
+- Status: Draft
+- Discussion: [#49693](https://github.com/pandas-dev/pandas/pull/49693)
+- Author: [Marco Gorelli](https://github.com/MarcoGorelli)
+- Revision: 1
+
+## Abstract
+
+The suggestion is to add a `mode.no_default_index` option which, if enabled,
+would ensure:
+- if a ``DataFrame`` / ``Series`` is created, then by default it won't have an ``Index``;
+- nobody will get an ``Index`` unless they ask for one - this would affect the default behaviour of ``groupby``, ``value_counts``, ``pivot_table``, and more.
+
+This option would not be the default. Users would need to explicitly opt-in to it, via ``pd.set_option('mode.no_default_index', True)``, via ``pd.option_context``, or via the ``PANDAS_NO_DEFAULT_INDEX`` environment variable.
+
+## Motivation and Scope
+
+The Index can be a source of confusion and frustration for pandas users. For example, let's consider the inputs
+
+```python
+In [37]: ser1 = df.groupby('sender')['amount'].sum()
+
+In [38]: ser2 = df.groupby('receiver')['amount'].sum()
+
+In [39]: ser1
+Out[39]:
+sender
+1    10
+2    15
+3    20
+5    25
+Name: amount, dtype: int64
+
+In [40]: ser2
+Out[40]:
+receiver
+1    10
+2    15
+3    20
+4    25
+Name: amount, dtype: int64
+```
+. Then:
+
+- it can be unexpected that summing `Series` with the same length (but different indices) produces `NaN`s in the result (https://stackoverflow.com/q/66094702/4451315):
+
+  ```python
+  In [41]: ser1 + ser2
+  Out[41]:
+  1    20.0
+  2    30.0
+  3    40.0
+  4     NaN
+  5     NaN
+  Name: amount, dtype: float64
+  ```
+
+- concatenation, even with `ignore_index=True`, still aligns on the index (https://github.com/pandas-dev/pandas/issues/25349):
```

Comment:
Couldn't these examples be just as simple if a `RangeIndex(0, len(df))` was used in place of `NoIndex`? I suppose the only advantage of `NoIndex` over `RangeIndex` would be when operating with one frame that is `NoIndex` while the other has some `Index`.  What would happen there?  Is the `NoIndex` treated like a NumPy ndarray?  Or are there no matches with the existing index and so all `NaN`?  Or does it matter which is the left and right hand of the operator?

Reply:
The issue is that `RangeIndex` isn't preserved under slicing. If you do
```python
ser = pd.Series([1,2,3])
ser[ser != 2]
```
then the index would be `Int64Index([0, 2], dtype='int64')`

With `NoIndex`, on the other hand, the `NoIndex` would be preserved:
```python
In [1]: pd.set_option('mode.no_default_index', True)

In [2]: ser = pd.Series([1, 2, 3])

In [3]: ser
Out[3]: 
1
2
3
dtype: int64

In [4]: ser.index
Out[4]: NoIndex(len=3)

In [5]: ser[ser!=2].index
Out[5]: NoIndex(len=2)
```

---

Alignment would require both to be `NoIndex` of the same length